rust/core-lib/src/view.rs

```diff
@@ -400,54 +402,82 @@ impl View {
-            let (last, rest) = self.sel_regions().split_last().unwrap();
-            let mut sel = Selection::new();
-            for &region in rest {
-                sel.add_region(region);
+    /// Finds the unit of text containing the given offset.
+    fn unit(
+        &self,
+        text: &Rope,
+        offset: usize,
+        granularity: SelectionGranularity,
+    ) -> (usize, usize) {
+        match granularity {
+            SelectionGranularity::Point => (offset, offset),
+            SelectionGranularity::Word => {
+                let mut word_cursor = WordCursor::new(text, offset);
+                word_cursor.select_word()
             }
-            // TODO: small nit, merged region should be backward if end < start.
-            // This could be done by explicitly overriding, or by tweaking the
-            // merge logic.
-            sel.add_region(SelRegion::new(last.start, offset));
-            sel
-        };
-        let range_start = sel.last().unwrap().start;
-        self.set_selection(text, sel);
-        self.start_drag(range_start, range_start, range_start, SelectionGranularity::Point, false);
+            SelectionGranularity::Line => {
+                let (line, _) = self.offset_to_line_col(text, offset);
+                (
+                    self.line_col_to_offset(text, line as usize, 0),
+                    self.line_col_to_offset(text, (line as usize) + 1, 0),
+                )
+            }
+        }
     }
 
-    /// Selects the given region and supports multi selection.
-    fn select_region(&mut self, text: &Rope, region: SelRegion, multi_select: bool) {
-        let mut selection = match multi_select {
+    /// Selects text with a certain granularity and supports multi_selection
+    fn select(
+        &mut self,
+        text: &Rope,
+        offset: usize,
+        granularity: SelectionGranularity,
+        multi: bool,
+    ) {
+        // If multi-select is enabled, toggle existing regions
+        if multi
+            && granularity == SelectionGranularity::Point
+            && self.deselect_at_offset(text, offset)
+        {
+            return;
+        }
+
+        let (min, max) = self.unit(text, offset, granularity);
+        let region = SelRegion::new(min, max);
+
+        let base_sel = match multi {
             true => self.selection.clone(),
             false => Selection::new(),
         };
-
+        let mut selection = base_sel.clone();
         selection.add_region(region);
         self.set_selection(text, selection);
+
+        self.drag_state = Some(DragState { base_sel, min, max, granularity });
     }
 
-    /// Selects an entire word and supports multi selection.
-    pub fn select_word(&mut self, text: &Rope, offset: usize, multi_select: bool) {
-        let (start, end) = {
-            let mut word_cursor = WordCursor::new(text, offset);
-            word_cursor.select_word()
+    /// Selects a specific range (eg. when the user performs SHIFT + click).
```

Comment:
I might say, 'expands an existing selection', for the first bit, instead? I find this RPC name confusing, generally, because double clicking a word is actually "selecting a range"; the thing that makes this distinct is that the result depends on the existing state.